go/engine/paperprovision.go

```diff
@@ -0,0 +1,272 @@
+		return nil
+	}
+
+	// need a session to continue to provision, login with paper sigKey
+	return e.G().LoginState().LoginWithKey(ctx.LoginContext, e.User, kp.sigKey, afterLogin)
+}
+
+func (e *PaperProvisionEngine) sendNotification() {
+	e.G().NotifyRouter.HandleLogin(string(e.G().Env.GetUsername()))
+}
+
+func (e *PaperProvisionEngine) SubConsumers() []libkb.UIConsumer {
+	return []libkb.UIConsumer{}
+}
+
+func (e *PaperProvisionEngine) Result() error {
+	return e.result
+}
+
+// copied from loginProvision
+func (e *PaperProvisionEngine) fetchLKS(ctx *Context, encKey libkb.GenericKey) error {
+	gen, clientLKS, err := fetchLKS(ctx, e.G(), encKey)
+	if err != nil {
+		return err
+	}
+	e.lks = libkb.NewLKSecWithClientHalf(clientLKS, gen, e.User.GetUID(), e.G())
+	return nil
+}
+
+// copied from loginProvision
+// makeDeviceKeysWithSigner creates device keys given a signing
+// key.
+func (e *PaperProvisionEngine) makeDeviceKeysWithSigner(ctx *Context, signer libkb.GenericKey) error {
+	args, err := e.makeDeviceWrapArgs(ctx)
+	if err != nil {
+		return err
+	}
+	args.Signer = signer
+	args.IsEldest = false // just to be explicit
+	args.EldestKID = e.User.GetEldestKID()
+
+	return e.makeDeviceKeys(ctx, args)
+}
+
+// copied from loginProvision
+// makeDeviceWrapArgs creates a base set of args for DeviceWrap.
+// It ensures that LKSec is created.  It also gets a new device
+// name for this device.
+func (e *PaperProvisionEngine) makeDeviceWrapArgs(ctx *Context) (*DeviceWrapArgs, error) {
+	if err := e.ensureLKSec(ctx); err != nil {
+		return nil, err
+	}
+
+	return &DeviceWrapArgs{
+		Me:         e.User,
+		DeviceName: e.DeviceName,
+		DeviceType: "desktop",
+		Lks:        e.lks,
+	}, nil
+}
+
+// copied from loginProvision
+// makeDeviceKeys uses DeviceWrap to generate device keys.
+func (e *PaperProvisionEngine) makeDeviceKeys(ctx *Context, args *DeviceWrapArgs) error {
+	eng := NewDeviceWrap(args, e.G())
+	if err := RunEngine(eng, ctx); err != nil {
+		return err
+	}
+
+	// TODO: do we need these?
+	//	e.signingKey = eng.SigningKey()
+	//	e.encryptionKey = eng.EncryptionKey()
+	return nil
+}
+
+// copied from loginProvision
+// ensureLKSec ensures we have LKSec for saving device keys.
+func (e *PaperProvisionEngine) ensureLKSec(ctx *Context) error {
+	if e.lks != nil {
+		return nil
+	}
+
+	pps, err := e.ppStream(ctx)
+	if err != nil {
+		return err
+	}
+
+	e.lks = libkb.NewLKSec(pps, e.User.GetUID(), e.G())
+	return nil
+}
+
+// copied from loginProvision
+// ppStream gets the passphrase stream, either cached or via
+// SecretUI.
+func (e *PaperProvisionEngine) ppStream(ctx *Context) (*libkb.PassphraseStream, error) {
+	if ctx.LoginContext != nil {
+		cached := ctx.LoginContext.PassphraseStreamCache()
+		if cached == nil {
+			return nil, errors.New("loginProvision: ppStream() -> nil PassphraseStreamCache")
+		}
+		return cached.PassphraseStream(), nil
+	}
+	return e.G().LoginState().GetPassphraseStreamForUser(ctx.SecretUI, e.User.GetName())
```

Comment:
does it make sense to falling back to the secret UI for paper provision, since the whole point is batch mode operation?